client/context/context.jsx

```diff
@@ -1,11 +1,240 @@
+        success: true
+      };
+    case 'JOIN_BOOKCLUB':
+      return {
+        ...state,
+        members: [...state.members, action.payload],
+        success: true
+      };
+    case 'LEAVE_BOOKCLUB':
+      return {
+        ...state,
+        members: state.members.filter((member) => member !== action.payload),
+        success: true
+      };
+    case 'SEARCH_BOOKCLUB':
+      return {
+        ...state,
+        club: action.payload,
+        success: true
+      };
+    case 'FUZZY_BOOKCLUB_SEARCH':
+      return {
+        ...state,
+        fuzzyClubs: action.payload,
+        success: true
+      };
+    case 'UPDATE_KEYWORD':
+      return {
+        ...state,
+        keyword: action.payload
+      };
+    case 'SEARCH_ERROR':
+      return {
+        ...state,
+        error: action.payload,
+        success: false
+      };
+    default:
+      return state;
+  }
+};
+
+export const AppContext = createContext(initState);
 
 export const AppProvider = ({ children }) => {
-  const [selectedClubData, setSelectedClubData] = useState({});
+  const [state, dispatch] = useReducer(reducer, initState);
+
+  useEffect(() => {
+    getClubById('602bff381017a68f02009b0e');
+    getClubs();
+  }, []);
+
+  // Actions
+  // Get all bookclubs
+  async function getClubs() {
+    try {
+      const res = await axios.get('/bookclub/all');
+
+      dispatch({
+        type: 'GET_BOOKCLUBS',
+        payload: res.data
+      });
+    } catch (err) {
+      dispatch({
+        type: 'SEARCH_ERROR',
+        payload: err.response.data.error
+      });
+    }
+  }
+
+  // Get single bookclub by name
+  async function getClubByName(name) {
+    try {
+      const res = await axios.get(`/`, { name });
+
+      dispatch({
+        type: 'GET_BOOKCLUB',
+        payload: res.data
+      });
+    } catch (err) {
+      dispatch({
+        type: 'SEARCH_ERROR',
+        payload: err.response.data.error
+      });
+    }
+  }
+
+  // Get single bookclub by id
+  async function getClubById(id) {
+    try {
+      const res = await axios.get(`/bookclub/${id}`);
+
+      dispatch({
+        type: 'SEARCH_BOOKCLUB',
+        payload: res.data
+      });
+    } catch (err) {
+      dispatch({
+        type: 'SEARCH_ERROR',
+        payload: err.response.data.error
+      });
+    }
+  }
+
+  // Join bookclub by id
+  async function joinClubById(id, userId) {
+    try {
+      await axios.post(`/bookclub/join/${id}`, {
+        userId
+      });
+
+      dispatch({
+        type: 'JOIN_BOOKCLUB',
+        payload: user._id
+      });
+    } catch (err) {
+      dispatch({
+        type: 'SEARCH_ERROR',
+        payload: err.response.data.error
+      });
+    }
+  }
+
+  // Leave bookclub by id
+  async function leaveClubById(id, userId) {
+    try {
+      await axios.post(`/bookclub/join/${id}`, {
+        userId
+      });
+
+      dispatch({
+        type: 'LEAVE_BOOKCLUB',
+        payload: user._id
+      });
```

Comment:
Oops wrong spot throwing the error. The above two functions both go to the catch block, and then throw an error in that catch block of trying to read 'data' of undefined

Reply:
Which one was this? Not using dispatch/reducer now. This one may not using the global state setup on Daniel's end, I think he was on join/leave functionality.